sympy/solvers/tests/test_solveset.py

```diff
@@ -1626,6 +1532,10 @@ def test_nonlinsolve_using_substitution():
     soln = (-s_x, y)
     assert nonlinsolve(system, [x, y]) == FiniteSet(soln)
 
+@XFAIL
+def test_nonlinsolve_using_substitution_1():
+    # giving incomplete solution due to not
+    # complete development of lambert
```

Comment:
Why does this fail?

Presumably the output of `solveset` has changed somehow. What case is `solveset` returning something different for?

Can you show the code that makes the equation and calls `solveset` and what the output is on master and with this PR?

Reply:
> what the output is on master and with this PR?
```
>>> from sympy import *
>>> from sympy.abc import x,y,z,a,b,c
>>> syms = [y, x, z]
>>> system = [z**2*x**2 - z**2*y**2/exp(x)]
>>> nonlinsolve(system,syms)
FiniteSet((y, x, 0), (y, 2*LambertW(-y/2), z), (y, 2*LambertW(y/2), z))
```
And this is not the complete solution!
The correct solution is  
```
FiniteSet((y, x, 0), \
        (y, ImageSet(Lambda(n, 2*LambertW(-y/2, n)), S.Integers), z), \
            (y, ImageSet(Lambda(n, 2*LambertW(y/2, n)), S.Integers), z))
```

Reply:
> Can you show the code that makes the equation and calls solveset 

https://github.com/sympy/sympy/blob/d192a8b8b90bf1ac15b5f06c1cd254d19ed7d328/sympy/solvers/solveset.py#L3372-L3385

Reply:
I'm asking what the input and output of `solveset` is (not `nonlinsolve`). Something like:
```
In [95]: eq = x**2 + 1

In [96]: solveset(eq, x, Reals)
Out[96]: ∅
```

Reply:
> I'm asking what the input and output of solveset is (not nonlinsolve). Something like:


```
>>> eq = z**2*x**2 - z**2*y**2/exp(x)
>>> solveset(eq, x, Reals)
Intersection(FiniteSet(2*LambertW(-y/2), 2*LambertW(y/2)), Reals)
```

Reply:
Are you sure it's over the `Reals` and not `Complexes`?

This PR gives:
```python
In [24]: solveset(eq, x, Complexes)
Out[24]: 
⎧   ⎛-y ⎞     ⎛y⎞⎫
⎨2⋅W⎜───⎟, 2⋅W⎜─⎟⎬
⎩   ⎝ 2 ⎠     ⎝2⎠⎭
```
(which is incorrect)

I thought that this PR was just for `solveset_real`. That should mean that the output over Complexes should not be changed.

Reply:
> I thought that this PR was just for solveset_real. That should mean that the output over Complexes should not be changed.

Yes, this PR gives the solutions for both `solveset_real` and `solveset`(complexes) and It returns wrong solutions for reals also.
e.g. (in XFAIL)

https://github.com/sympy/sympy/blob/d192a8b8b90bf1ac15b5f06c1cd254d19ed7d328/sympy/solvers/tests/test_solveset.py#L2300-L2318

Reply:
I think that this PR should be changed so that it

1. Does not attempt to solve over Complexes.
2. Does not return incorrect solutions over Reals.

Reply:
> I think that this PR should be changed so that it
> 
> Does not attempt to solve over Complexes.
> Does not return incorrect solutions over Reals.

It takes a lot of code to do so!

Reply:
> I think that this PR should be changed so that it
> 
> Does not attempt to solve over Complexes.

Oh actually after removing the above code from `_invert_complex`, it's not returning code for complexes!
The first point is achieved :)

Reply:
Does that mean that there is no change to this test for `nonlinsolve`?

Reply:
> Does that mean that there is no change to this test for nonlinsolve?



```
>>> system = [z**2*x**2 - z**2*y**2/exp(x)]
>>> syms = [y, x, z]
>>> nonlinsolve(system,syms) 
FiniteSet((y, x, 0), (y, 2*LambertW(-y/2), z), (y, 2*LambertW(y/2), z), (x*sqrt(exp(x)), x, z), (-x*sqrt(exp(x)), x, z))
```
(not a complete solution)
This is the change since this collects both the equations (of `solveset` i.e. `real` and `complexes`).

https://github.com/sympy/sympy/blob/d192a8b8b90bf1ac15b5f06c1cd254d19ed7d328/sympy/solvers/solveset.py#L3372-L3384

Reply:
Please give your opinion on this test case .

Reply:
I think I preferred the previous output but I can see that both are equally "correct" so I guess this is okay.

Reply:
Actually I don't think that they are equally correct. Looking at the equation it is clear that the solutions should be `z = 0` and `y = +- sqrt(x**2*exp(x))` which is what was being returned before.

Inverting that second equation to `x = 2*LambertW(-y/2)` is not correct because the full solution for `x` should be `x = 2*LambertW(-y/2, n)` for any integer `n`. Ideally `nonlinsolve` would not try to invert this because in the equation `y**2 = x**2 * exp(x)` it's much better to invert for `y` than for `x`. Also we can invert for `y` over the complex numbers more easily whereas inverting for `x` here only works over the reals.

Is this just a bug in `nonlinsolve`?

Why doesn't this happen in other cases e.g.:
```python
In [5]: nonlinsolve([sin(x) - y**2], [x, y])                                                                                      
Out[5]: 
⎧⎛      ________⎞  ⎛     ________⎞⎫
⎨⎝x, -╲╱ sin(x) ⎠, ⎝x, ╲╱ sin(x) ⎠⎬
⎩                                 ⎭
```

Reply:
Here, we should understand that `linsolve` is trying to add up both `complex` and `real` solutions (in `L-3388` below). As we know that `lambert` is only developed for `reals` but not over `complexes` so it's giving the lambert solutions for just `reals` but not `complexes` hence we getting the solution in that way.

https://github.com/sympy/sympy/blob/d192a8b8b90bf1ac15b5f06c1cd254d19ed7d328/sympy/solvers/solveset.py#L3372-L3390

Reply:
> Ideally nonlinsolve would not try to invert this because in the equation y**2 = x**2 * exp(x) it's much better to invert for y than for x. Also we can invert for y over the complex numbers more easily whereas inverting for x here only works over the reals.
> 
> Is this just a bug in nonlinsolve?
> 
> Why doesn't this happen in other cases e.g.:

I don't think it's a bug, as you may see in some other cases also the pattern of returning solutions is the similar.(i.e. the solutions is in terms of `y` instead of `x`).
e.g.
```
>>> nonlinsolve([(x+y)**2 - 4, x + y - 2], [x, y])
FiniteSet((2 - y, y))

>>> nonlinsolve([(x+y)**2 - z*4, z*x + y - 2], [x, y,z])
FiniteSet((2 - y, y, 1), (-(-2 + (-2*sqrt(z**3) - 2)/(z - 1))/z, (-2*sqrt(z**3) - 2)/(z - 1), Complement(FiniteSet(z), FiniteSet(0))), (-(-2 + (2*sqrt(z**3) - 2)/(z - 1))/z, (2*sqrt(z**3) - 2)/(z - 1), Complement(FiniteSet(z), FiniteSet(0))))

or 

⎧⎛ ⎛            ____    ⎞                             ⎞  ⎛ ⎛          ____    ⎞                           ⎞               ⎫
⎪⎜ ⎜           ╱  3     ⎟                             ⎟  ⎜ ⎜         ╱  3     ⎟                           ⎟               ⎪
⎪⎜ ⎜     - 2⋅╲╱  z   - 2⎟          ____               ⎟  ⎜ ⎜     2⋅╲╱  z   - 2⎟        ____               ⎟               ⎪
⎨⎜-⎜-2 + ───────────────⎟         ╱  3                ⎟  ⎜-⎜-2 + ─────────────⎟       ╱  3                ⎟               ⎬
⎪⎜ ⎝          z - 1     ⎠   - 2⋅╲╱  z   - 2           ⎟  ⎜ ⎝         z - 1    ⎠   2⋅╲╱  z   - 2           ⎟               ⎪
⎪⎜────────────────────────, ───────────────, {z} \ {0}⎟, ⎜──────────────────────, ─────────────, {z} \ {0}⎟, (2 - y, y, 1)⎪
⎩⎝           z                   z - 1                ⎠  ⎝          z                 z - 1               ⎠               ⎭
```

Reply:
I hope you've got an Idea .